### PR TITLE
Save & restore workspace layout on navigation

### LINKED ISSUE #1278: subscriptions: prevent multiple subscribe/unsubscribe events
Currently we start subscriptions on a per-component basis, which creates the potential for problems where multiple subscribing components are involved.

E.G. if there are two components the sequence of events might look like this:

1. Component 1 initialises and registers its subscription.
2. The subscription is started.
3. Component 2 initialises and registers its subscription.
4. The subscription is stopped and re-started with the recomputed query.

We need to change this order of events to:

1. Component 1 initialises and registers its subscription.
2. Component 2 initialises and registers its subscription.
3. The subscription is started.

This bug doesn't bite at the moment because we only initialise the workspace with one component, however with #662 that will change.

See https://github.com/cylc/cylc-ui/pull/1249#discussion_r1119106019_ for more details

### PR DESCRIPTION
**Pull request by MetRonnie**
Closes #1278
Partially addresses #662

> 2. Save the layout last used by a workspace (important).
>     * When the user navigates away from a workflow, save the tab layout.
>     * When the user navigates back to the workflow, load the tab layout.
>     * When the user navigates to a new run of the same workflow, copy the layout from the last run of the workflow.

The layout is saved in the vuex store so only persists for the browser session. Ongoing investigation into saving it in local storage.

demo14

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Tests are included
- [x] `CHANGES.md` entry included
- [ ] Changelog entry added in cylc-doc

### CODE REVIEW

File: src/router/index.js [hunk 1]
```diff
@@ -76,9 +76,7 @@ router.beforeEach(async (to, from) => {
   if (!store.state.user.user.permissions.includes('read') && to.name !== 'noAuth') {
     return { name: 'noAuth' }
   }
-})
 
-router.beforeResolve((to, from) => {
```
Comment (oliver-sanders) [inline src/router/index.js:81]: Interesting, I would like to understand this change.

File: src/store/app.module.js [hunk 2]
```diff
@@ -27,6 +30,16 @@ const mutations = {
   setTitle (state, title) {
     state.title = title
   },
+  saveLayout (state, { workflowName, layout, views }) {
+    /* NOTE: use markRaw to prevent proxying of the Lumino layout in particular.
+    It is not necessary for this saved state to be reactive, and moreover
+    proxying the layout breaks some parts of the 3rd party Lumino backend. */
+    state.workspaceLayouts.set(workflowName, markRaw({ layout, views }))
+    if (state.workspaceLayouts.size > 10) {
```
Comment (oliver-sanders) [inline src/store/app.module.js:38]: Would be worth building a really complex layout and finding out how heavy it is.

I'm hoping that we can safely raise this number.

**Review by oliver-sanders — COMMENTED** [review]
Slick.

File: src/router/index.js (hunk 1 above)
Comment (MetRonnie) [reply src/router/index.js:81]: Discussed in person. When navigating from one workflow to another, the subscription would start using the old workflow name for a short while and then update to the new workflow name. Now the subscription starts using the new workflow name

**Comment by MetRonnie** [comment]
Drafting while I see if I can fix #1278 before we merge this

**Comment by MetRonnie** [comment]
@oliver-sanders and @markgrahamdawson I've opened a secondary PR against this one that modifies the log view's `intialOptions` prop to allow it to store the view state. Mark, this approach can be used on the other views but isn't implemented yet. https://github.com/MetRonnie/cylc-ui/pull/7

File: src/components/cylc/workflow/Lumino.vue [hunk 3]
```diff
@@ -15,187 +15,243 @@ You should have received a copy of the GNU General Public License
...
+const views = ref(new Map())
+
+const defaultView = useDefaultView()
+
+// create a box panel, which holds the dock panel, and controls its layout
+const boxPanel = new BoxPanel({ direction: 'left-to-right', spacing: 0 })
+// create dock panel, which holds the widgets
+const dockPanel = new DockPanel()
+boxPanel.addWidget(dockPanel)
+BoxPanel.setStretch(dockPanel, 1)
```
Comment (oliver-sanders) [inline src/components/cylc/workflow/Lumino.vue:111]: Is that meant to be `BoxPanel.setStretch` with a capital `B`?

File: src/components/cylc/workflow/Lumino.vue [hunk 4]
```diff
@@ -15,187 +15,243 @@ You should have received a copy of the GNU General Public License
...
-     */
-    syncWidgets (newVal, oldVal) {
-      for (const [id, item] of Object.entries(newVal)) {
-        if (!(id in oldVal)) {
-          this.addWidget(id, item.view)
-        }
+/**
+ * Remove all the widgets present in the DockPanel.
+ */
+const closeAllViews = () => {
```
Comment (oliver-sanders) [inline src/components/cylc/workflow/Lumino.vue:158]: IMO:

```js
function foo(bar) { ... }
```

Is cleaner than

```js
const foo = (bar) => { ... }
```

What's the motive for the `const` approach? Does it do something different?

File: src/components/cylc/workflow/Lumino.vue (hunk 3 above)
Comment (MetRonnie) [reply src/components/cylc/workflow/Lumino.vue:111]: Yes, same as before

File: src/components/cylc/workflow/Lumino.vue (hunk 4 above)
Comment (MetRonnie) [reply src/components/cylc/workflow/Lumino.vue:158]: No preference from me

Comment (oliver-sanders) [reply src/components/cylc/workflow/Lumino.vue:158]: So the only functional difference is the implicit `this` variable?

Comment (MetRonnie) [reply src/components/cylc/workflow/Lumino.vue:158]: Well, also declarations (`function foo () {}`) are hoisted but expressions (`const foo = () => {}`) are not.

**Pull request merged**